### DISCUSSION TITLE
Why are binding redirects always necessary?

### DISCUSSION THREAD

**Discussion by chuchu** [dotnet/runtime#46199] [General]
Hi All.
I hope this question was not already answered here. I could not find it.
Why do the versions of dependencies, listed in a nuspec file, not match the concrete referenced version? For instance the nuget package for Microsoft.Extensions.Configuration version 3.1.1 lists a dependency to Microsoft.Extensions.Configuration.Abstractions 3.1.1 in its nuspec file. But the Microsoft.Extensions.Configuration.dll in version 3.1.1 references Microsoft.Extensions.Configuration.Abstractions 3.1.**0**. Having it this way it is mandatory to use binding redirects. What is the reason for this?

**Comment by huoyaoyuan** [comment]
#46182 ?

**Reply by chuchu** [reply]
Very similar, but I'm not sure if it is the same. For me, the above mentioned behavior looks like an explicit design decision. Each assembly of net core references always other net core assemblies with the patch version "0". Even when the nuspec states something different. Take for example the **Microsoft.Extensions.Configuration** package in version **3.1.1**. The nuspec contains:
```xml
<dependencies>
  <group targetFramework=".NETCoreApp3.1">
    <dependency id="Microsoft.Extensions.Configuration.Abstractions" version="3.1.1" exclude="Build,Analyzers" />
  </group>
  <group targetFramework=".NETStandard2.0">
    <dependency id="Microsoft.Extensions.Configuration.Abstractions" version="3.1.1" exclude="Build,Analyzers" />
  </group>
</dependencies>
```
But the actual reference in the assembly points no to Microsoft.Extensions.Configuration.Abstractions 3.1.1 instead of 3.1.0.
![image](https://user-images.githubusercontent.com/521666/102988368-1b066b00-4514-11eb-9a8f-44e1cb73d02a.png)
This is resolved with a binding redirects in the app.config:
```xml
<dependentAssembly>
  <assemblyIdentity name="Microsoft.Extensions.Configuration.Abstractions" publicKeyToken="adb9793829ddae60" culture="neutral" />
  <bindingRedirect oldVersion="0.0.0.0-3.1.1.0" newVersion="3.1.1.0" />
</dependentAssembly>
```
I would like to understand why the real assembly does not reference the Abstractions assembly in version 3.1.1 directly rendering the binding redirect unnecessary.

**Comment by gaborposz** [comment]
Hello,
For us it's not just annoying, but also causing a lot of extra work.

We are using these packages as "external software": Not as NuGet packages, but simply as assemblies, deployed directly with our applications and other assemblies in one huge global bin folder.

Therefore [automatic binding redirection generation](https://docs.microsoft.com/en-us/dotnet/framework/configure-apps/how-to-enable-and-disable-automatic-binding-redirection) does not work for us, we have to do it manually. And it is quite cumbersome if you have dozens of app.config files, as well as a lot of "external assemblies" :(

Of course it would be easier if we could use them as NuGet packages with automatic binding redirection generation, but it's a very big, 10+ years old project that should - by the way - go on for another 10+ years, and we can't change such basic principles anymore.

So it would make our life much easier if the versions are referenced correctly in all the assemblies.
